File: 0046_Permutations/Permutations.java
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> ans = new ArrayList<List<Integer>>();
        List<Integer> curr = new ArrayList<Integer>();
        boolean[] used = new boolean[nums.length];
        DFS(nums, 0, used, curr, ans);
        return ans;
    }
    
    private void DFS(int[] nums, int s, boolean[] used, List<Integer> curr, List<List<Integer>> ans){
        if(curr.size() == nums.length){
            ans.add(new ArrayList<Integer>(curr));
            return;
        }
        
        for(int i=0;i<nums.length;i++){
            if(used[i]) continue;
            curr.add(nums[i]);
            used[i] = true;
            DFS(nums, i+1, used, curr, ans);
            curr.remove(curr.size()-1);
            used[i] = false;
        }
    }
}

File: 0046_Permutations/Permutations.cs
public class Solution {
    public IList<IList<int>> Permute(int[] nums) {
        var ans = new List<IList<int>>();
        if(nums == null || nums.Length == 0) return ans;
        var used = new bool[nums.Length];
        DFS(nums, used, new List<int>(), ans);
        return ans;
    }
    
    private void DFS(int[] nums, bool[] used, IList<int> curr, IList<IList<int>> ans){
        if(curr.Count == nums.Length){
            ans.Add(new List<int>(curr));
            return;
        }
        
        for(int i=0;i<nums.Length;i++){
            if(used[i]) continue;
            used[i] = true;
            curr.Add(nums[i]);
            DFS(nums, used, curr, ans);
            used[i] = false;
            curr.RemoveAt(curr.Count - 1);
        }
    }
}